### PR TITLE
No gas and support for Request Listings and address verifiers

### PR DESCRIPTION
**Pull request by crazybuster**
First pull request? Read our [guide to contributing](http://docs.originprotocol.com/#contributing)

### Description:

Please explain the changes you made here:

- Allow for centralized no gas creation
- Support for request listings
- Support for verifier

### Checklist:

- [ ] Test your work and double-check to confirm that you didn't break anything
- [ ] Wrap any displayed ETH addresses with [`formattedAddress`](https://github.com/OriginProtocol/origin/blob/master/origin-dapp/src/utils/user.js#L15-L17)
- [ ] Wrap any new text/strings for translation
- [ ] Run `npm run translations` if there are any changes to translated strings
- [ ] Map any new environment variables with a default value in the Webpack config
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/origin-docs)

### CODE REVIEW

File: origin-discovery/src/apollo/injector.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  console.log('creator vs seller:', listing.creator, listing.seller)
+  if (listing.creator != listing.seller)
+  {
+    throw('Creator must be same as the seller!')
+  }
+  //looks like I need a raw response to verify this hash else it hashes the processed one
+
+  if (!(await origin.marketplace.verifyListingSignature(listing, listing.seller)))
+  {
+    throw('Signature does not match that of the seller')
```
Comment (franckc) [inline origin-discovery/src/apollo/injector.js]: nit: i think we tend to use throw new Error(<message>)

File: origin-discovery/src/listener/handler_marketplace.js [hunk 2]
```diff
@@ -312,4 +339,73 @@ class MarketplaceEventHandler {
...
+   * @returns {Promise<{listing: Listing, offer: Offer, seller: User, buyer: User}>}
+   * @private
+   */
+  async _getOfferDetails(log, origin, blockInfo) {
+    const listingId = generateNoGasListingId(log)
+    const offerId = generateNoGasOfferId(log)
+
+    const offer = await origin.marketplace.getOffer(offerId)
+
+    console.log('offer details', offer)
```
Comment (franckc) [inline origin-discovery/src/listener/handler_marketplace.js]: nit: use logger.info or logger.debug

File: origin-discovery/src/listener/handler_marketplace.js [hunk 3]
```diff
@@ -312,4 +339,73 @@ class MarketplaceEventHandler {
...
+    checkEventsFreshness(offer.events, blockInfo)
+
+    // TODO: need to load from db to verify that the listingIpfs haven't already been set!!!
+    //const status = web3.utils.toBN(offer.seller) != 0 ? 'pending': 'active'
```
Comment (franckc) [inline origin-discovery/src/listener/handler_marketplace.js]: remove ?

File: origin-js/src/index.js [hunk 4]
```diff
@@ -72,11 +75,12 @@ export default class Origin {
         contractService: this.contractService,
         discoveryService: this.discoveryService,
         ipfsService: this.ipfsService,
+        hotService: this.hotService,
         affiliate,
         arbitrator,
         store,
         blockEpoch,
-        perfModeEnabled
+        perfModeEnabled,
```
Comment (franckc) [inline origin-js/src/index.js]: extra semi column may break lint ?

File: origin-contracts/contracts/marketplace/vA/Marketplace.sol [hunk 5]
```diff
@@ -0,0 +1,477 @@
+pragma solidity ^0.4.24;
+
+import "../../../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+/**
+ * @title A Marketplace contract for managing listings, offers, payments, escrow and arbitration
+ * @author Nick Poulden <nick@poulden.com>
```
Comment (franckc) [inline origin-contracts/contracts/marketplace/vA/Marketplace.sol]: want to put your name ?

Comment (tomlinton) [reply origin-contracts/contracts/marketplace/vA/Marketplace.sol]: Who is this Nick guy anyway? 😁

Comment (crazybuster) [reply origin-contracts/contracts/marketplace/vA/Marketplace.sol]: He's the author

**Pull request merged**